(SAVE FOR 2.6 PLANNING SPRINT) Add GHA to test AsciiDoc builds on each PR

dfarrell07 opened the pull request:
Add a GitHub action that builds each .adoc file on each PR, failing if
there are any errors WARN-level or above. All issues are shown in the
logs before failing.

Subsequent PRs could lower the threshold for failing to INFO and fix
additional issues.

This could eventually be extracted to a stand-alone GitHub action,
perhaps moved to a different repo, and shared with the rest of the
AsciiDoc ecosystem.

Shared automated testing infra makes the contribution process more
comfortable for developers and the code review process easier for
owners. It also guarantees consistent standards across the project.

Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

dfarrell07 committed:
Add GHA to test AsciiDoc builds on each PR

Add a GitHub action that builds each .adoc file on each PR, failing if
there are any errors WARN-level or above. All issues are shown in the
logs before failing.

Subsequent PRs could lower the threshold for failing to INFO and fix
additional issues.

This could eventually be extracted to a stand-alone GitHub action,
perhaps moved to a different repo, and shared with the rest of the
AsciiDoc ecosystem.

Shared automated testing infra makes the contribution process more
comfortable for developers and the code review process easier for
owners. It also guarantees consistent standards across the project.

Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

dfarrell07 commented:
/cc @skitt 

Mind reviewing? :nerd_face: :pray:

dfarrell07 commented:
Here are examples of correct-failures when the threshold is lowered:

https://github.com/dfarrell07/rhacm-docs/pull/2
https://github.com/dfarrell07/rhacm-docs/pull/3

dfarrell07 commented:
Here are the things Asciidoctor will throw warnings about on (it's not as strict as I hoped, but okay): https://docs.asciidoctor.org/asciidoctor/latest/errors-and-warnings/

skitt commented:
This seems rather useful, especially the detection of missing cross-references! Next step, Prow ;-).

/approve

swopebe commented:
Doc team is not sure what this is? Can you point to an issue or BZ for us, or JIRA if this is to correct a build issue that we are unaware of???

Thank you!

dfarrell07 commented:
Hello @swopebe! This is discussed on https://github.com/stolostron/backlog/issues/19403#issuecomment-1024419675, and the what/why are documented in the commit message. This is a "drive-thru contribution" of CI for AsciiDocs. VM Brasseur (who I think works for Red Hat now?) has a great [write-up about contributions like this](https://www.fosslife.org/open-source-drive-thru-contributors). As I mentioned on your template UX call last Friday, @skitt and I are also in the queue to host an ["open source best practices" Lunch and Learn](https://docs.google.com/document/d/1Y_S1oOpd9rMSm4sGyg5byWjvj7dXlvT-Ao38yGk0xr8) that would be a great venue to talk about this way of working at Red Hat if you're interested.

swopebe commented:
Ok. Will look when I get a chance.

We use the **Portal stage for the docs**, and we try to push that as a way to see the build after the merge, but I think what you are trying to do is similar **to the tech preview docs I have here:** https://github.com/stolostron/mce-docs (see the github workflow)

But that is a temporary place.

 Red Hat pushed us to use the Customer Portal and Customer Portal stage when we moved here--as this is the larger doc strategy. 

So long as this process for our team and larger engineers remain the way we have it, a workflow in your personal directory could work. We worry about having another stage environment outside of the portal to maintain or discuss with the engineering team. The docs are built and located in the Portal, so we try to show it there (as all the products in the portal do).

dfarrell07 commented:
Thanks for the pointers, I'll dig into those. I have a lot to learn about how the docs automation works.

The goal here isn't to create a new staging environment, but to catch bugs in AsciiDoc syntax before they're merged at all. This automation just does a build on each .adoc file and fails if any of the builds fails. The built files are thrown away at the end of the job. I set the failure threshold at a forgiving level to start, but plan to lower it and fix errors it identifies in current .adoc files in subsequent PRs. It's a different level of testing than the staging environment reviews by humans. Ideally, this would mostly be used by contributors to catch and fix errors without any reviewer work at all.

swopebe commented:
Yes, I understand the goal completely; we have a build system for that and we work within the confines of Pantheon and the Portal. What we are hoping to avoid is another environment for our developers that will steer everyone away from looking at the portal stage....our doc team gets reviews on the PRs to avoid build errors and navigation errors. 

That said, if you do this in your own fork, that is great.

dfarrell07 commented:
> our doc team gets reviews on the PRs to avoid build errors and navigation errors.

Maybe an example here will help clarify how this is applies in your case. As I mentioned before, https://github.com/dfarrell07/rhacm-docs/pull/2 and https://github.com/dfarrell07/rhacm-docs/pull/3 show examples of bugs that have passed review and are currently in the codebase. Using `git blame` on the first issue identified, I see it was introduced in https://github.com/stolostron/rhacm-docs/pull/3090. A heading was removed without de-indenting child headings. The PR did not receive non-author code review, but would have received this feedback from the automation:

```
asciidoctor: WARNING: cluster_pool_manage.adoc: line 22: section title out of sequence: expected level 3, got level 4
asciidoctor: WARNING: cluster_pool_manage.adoc: line 46: section title out of sequence: expected level 3, got level 4
```

dfarrell07 commented:
I modified the GHA in https://github.com/dfarrell07/rhacm-docs/pull/4 to test all files before failing the job if there are any warnings. That flags other examples of how CI can supplement code review and visual inspection. For example, https://github.com/stolostron/rhacm-docs/pull/2192 / https://github.com/stolostron/rhacm-docs/pull/2192/commits/489e0329219e6d1f3f96c3f5d782bc90d33fcf7c also didn't receive non-author code review and introduced a bug (duplicate reference) that would have been stopped by CI.

```
asciidoctor: WARNING: trouble_agent_credentials.adoc: line 15: id assigned to section already in use: 
```

swopebe commented:
When we merge, we build and we have an error check in Pantheon--just like this one. This is also why we have peer reviews where we should be checking this stuff. 

I am also using this time to remind the team that we need to do peer reviews @cadawson. 

Please feel free to use your process in your repo if that helps, though.

dfarrell07 commented:
> I am also using this time to remind the team that we need to do peer reviews @cadawson.

I don't think the issue leading to bugs being in the codebase is the team's reviews, as the PRs were only open for 4m and 3.5h.

I feel we're at a bit of an unfortunate impasse. I'd just say again that CI is absolutely required for developing quality software at scale and is a key goal of Red Hat. I'd be happy to discuss this more in a call, maybe with some other more senior engineers/docs folks, if you're interested.

swopebe commented:
@dfarrell07 I understand what you are saying, I am actually the one on the team that launched an open source process from the ground up at IBM (Open Liberty), the problem is we are a small team and we are committed to the Portal and have already a process in place and not enough resources to change that for the team.

The unfortunate event leading to errors in the code is another example of **not using the peer review process** that we have also committed to. This is something I have already raised to my former manager and will continue to raise. 

We use: Issues from developers, write our doc in ascii, merge into the source after dev and (hopefully) peer review (clearly this was not consistent), then build in Pantheon, that is our process at the moment. I am voting to not change this process mid-release, we can chat in a few weeks when we are not so behind if that works for you.

dfarrell07 commented:
> I am actually the one on the team that launched an open source process from the ground up at IBM (Open Liberty)

Very cool! I hope we can work together to keep innovating in Open Source going forward!

> I am voting to not change this process mid-release, we can chat in a few weeks when we are not so behind if that works for you.

Understand, that make a lot of sense. I'd be happy to triage this as needing more discussion in your next cycle's planning. :+1:

swopebe commented:
Yes, tomorrow I will open a "future" issue to explore using ascii action to preview the doc, then we can test it on the repo (I did this for MCE) and open it up to the team if the doc team of 3 agrees!

dfarrell07 commented:
I refactored this to show all the errors before failing and lowered the failing threshold to WARNING as the actual errors at that level do seem significant.

There are a number of issues identified in the current code. See the "Details" link in the "Linting / Build AsciiDocs" GitHub Action.

```
Bulding troubleshooting/trouble_agent_credentials.adoc
asciidoctor: WARNING: trouble_agent_credentials.adoc: line 15: id assigned to section already in use: troubleshooting-cluster-rotating-agents
```

```
Bulding clusters/master.adoc
asciidoctor: ERROR: master.adoc: line 26: include file not found: /rhacm-docs/clusters/man_clust_taints_tol.adoc
asciidoctor: ERROR: clusterclaims.adoc: line 2: level 0 sections can only be used when doctype is book
asciidoctor: WARNING: list_clusterclaim.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: custom_clusterclaims.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: upgrade_cluster.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: remove_managed_cluster.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: backup_and_restore.adoc: line 2: section title out of sequence: expected level 1, got level 2
Bulding clusters/main.adoc
asciidoctor: ERROR: main.adoc: line 26: include file not found: /rhacm-docs/clusters/man_clust_taints_tol.adoc
asciidoctor: ERROR: clusterclaims.adoc: line 2: level 0 sections can only be used when doctype is book
asciidoctor: WARNING: list_clusterclaim.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: custom_clusterclaims.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: upgrade_cluster.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: remove_managed_cluster.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: backup_and_restore.adoc: line 2: section title out of sequence: expected level 1, got level 2
```

```
Bulding multicluster_engine/master.adoc
asciidoctor: ERROR: master.adoc: line 45: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_manage.adoc
asciidoctor: ERROR: master.adoc: line 46: include file not found: /rhacm-docs/multicluster_engine/cluster_claim.adoc
asciidoctor: ERROR: master.adoc: line 47: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_scale.adoc
asciidoctor: ERROR: master.adoc: line 48: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_rel_img_update.adoc
asciidoctor: ERROR: master.adoc: line 49: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_destroy.adoc
asciidoctor: ERROR: hosted_control_planes_intro.adoc: line 2: level 0 sections can only be used when doctype is book
asciidoctor: WARNING: hosted_control_planes_configure.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: disable_hosted_control_planes.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: discovery_intro.adoc: line 2: section title out of sequence: expected level 1, got level 2
Bulding multicluster_engine/create_azure.adoc
Bulding multicluster_engine/install_connected.adoc
Bulding multicluster_engine/main.adoc
asciidoctor: ERROR: main.adoc: line 45: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_manage.adoc
asciidoctor: ERROR: main.adoc: line 46: include file not found: /rhacm-docs/multicluster_engine/cluster_claim.adoc
asciidoctor: ERROR: main.adoc: line 47: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_scale.adoc
asciidoctor: ERROR: main.adoc: line 48: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_rel_img_update.adoc
asciidoctor: ERROR: main.adoc: line 49: include file not found: /rhacm-docs/multicluster_engine/cluster_pool_destroy.adoc
asciidoctor: ERROR: hosted_control_planes_intro.adoc: line 2: level 0 sections can only be used when doctype is book
asciidoctor: WARNING: hosted_control_planes_configure.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: disable_hosted_control_planes.adoc: line 2: section title out of sequence: expected level 1, got level 3
asciidoctor: WARNING: discovery_intro.adoc: line 2: section title out of sequence: expected level 1, got level 2
```

```
Bulding networking/master.adoc
asciidoctor: WARNING: network_config_hub.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: network_config_managed.adoc: line 2: section title out of sequence: expected level 1, got level 2
Bulding networking/main.adoc
asciidoctor: WARNING: network_config_hub.adoc: line 2: section title out of sequence: expected level 1, got level 2
asciidoctor: WARNING: network_config_managed.adoc: line 2: section title out of sequence: expected level 1, got level 2
Build errors were detected in 7 file(s), see per-file details above.
```

cadawson commented:
Thanks, @dfarrell07 . The timing of the run coincided with a big move of files from the clusters folder. I fixed a couple of the issues that looked like they were legitimate. The others, I think will be resolved now that the files are moved and stabilized, somewhat. Can we run the scan again to see a clearer picture? Thanks!

dfarrell07 commented:
> Can we run the scan again to see a clearer picture? Thanks!

Yep, the best way to test again with new changes is just to rebase the PR.

I rebased this PR. There are quite a few failures, too many to copy here.

https://github.com/stolostron/rhacm-docs/runs/7839465283?check_suite_focus=true

> Build errors were detected in 9 file(s), see per-file details above.

dfarrell07 commented:
/cc @skitt for a re-review on the substantially-changed bash.

dfarrell07 commented:
> substantially-changed bash.

The version you reviewed was like https://github.com/dfarrell07/rhacm-docs/commit/65af92ddefe0db4d51124b1fe01052aeafbe4777 FYI.

skitt commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,26 @@
+      - name: Build each .adoc with Asciidoctor, fail on WARN or above
+        run: |
+          FAILED=0
+          for doc in $(find $(pwd) -name '*.adoc' -printf "%P\n"); do
+            echo "Building $doc"
+            if ! docker run -v $(pwd):/rhacm-docs/ asciidoctor/docker-asciidoctor@sha256:59a7e165ed6375dd5dfb930cb90019f1ccaee39787dc97a3918d7022a7503543 asciidoctor --failure-level WARN --trace --verbose --warnings /rhacm-docs/$doc; then
+                FAILED=$((FAILED+1))
+            fi
+          done
+          if (( FAILED > 0 )); then
+            echo "Build errors were detected in $FAILED file(s), see per-file details above."
+            exit 1
+          fi
```
Parsing `find`’s output in this way is error-prone, and the overall work to be done can be greatly reduced by running `find` itself inside the container, and maximising the number of files given to each `asciidoctor` invocation:

```suggestion
          if ! docker run --rm -v $(pwd):/rhacm-docs/ asciidoctor/docker-asciidoctor@sha256:59a7e165ed6375dd5dfb930cb90019f1ccaee39787dc97a3918d7022a7503543 find /rhacm-docs/ -name \*.adoc -exec asciidoctor --failure-level ERROR --trace --verbose --warnings {} +; then
            echo "Build errors were detected, see per-file details above."
            exit 1
          fi
```

dfarrell07 replied:
Yes doing it all in one container invocation is much better, thanks! I changed the failure threshold back to WARN in the commited version FYI.

dfarrell07 commented:
Rebased, some smaller set of issues are still flagged.

dfarrell07 committed:
Only invoke asciidoctor container once

This greatly reduces the amount of work and is more copy+paste-able.

Co-authored-by: Stephen Kitt <steve@sk2.org>
Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

dfarrell07 commented:
I noticed the issues with broken links in 2.6 and was looking around again for link linting options that would work for AsciiDocs. We might be able to use one like this on the build HTML: https://github.com/stevenvachon/broken-link-checker
